cpp/src/arrow/memory_pool.cc

```diff
@@ -536,6 +543,39 @@ int64_t MemoryPool::max_memory() const { return -1; }
 // MemoryPool implementation that delegates its core duty
 // to an Allocator class.
 
+class ImmutableZeros : public Buffer {
+ public:
+  explicit ImmutableZeros(uint8_t* data, int64_t size, MemoryPool* pool)
+      : Buffer(data, size, CPUDevice::immutable_zeros_memory_manager(pool)),
+        pool_(pool) {}
+
+  ImmutableZeros() : Buffer(nullptr, 0), pool_(nullptr) {}
+
+  ~ImmutableZeros() override;
+
+  // Prevent copies and handle moves explicitly to avoid double free
```

Comment:
It sounds hackish to have a separate ImmutableZeros class, IMHO.

Reply:
I'm not sure why or what you mean by that? It's analogous to `PoolBuffer` in that it uses RAII to free memory when it is no longer needed, except two `Buffer` classes are needed to correctly model shared ownership, i.e. `ImmutableZeros` models the data, `ImmutableZerosPoolBuffer` models a shared reference to that data (even in contexts where a `unique_ptr` or raw pointer/reference to a `Buffer` is needed). Writing it down like this though, the names could use some refactoring, especially now that they both implement `Buffer`.